rust/src/lib/ifaces/inter_ifaces.rs

```diff
@@ -1017,6 +1019,106 @@ impl MergedInterfaces {
+                return Some(hsr_iface);
+            }
+
+            None
+        }) {
+            if let Some(hsr_conf) = &hsr_iface.hsr {
+                let mac_a =
+                    self.kernel_ifaces.get(&hsr_conf.port1).and_then(|m| {
+                        m.desired.as_ref().map(|iface| {
+                            iface.base_iface().mac_address.as_deref()
+                        })
+                    });
+                let mac_b =
+                    self.kernel_ifaces.get(&hsr_conf.port2).and_then(|m| {
+                        m.desired.as_ref().map(|iface| {
+                            iface.base_iface().mac_address.as_deref()
+                        })
+                    });
+
+                if mac_a.is_some() && mac_b.is_some() && mac_a != mac_b {
+                    return Err(NmstateError::new(
+                        ErrorKind::InvalidArgument,
+                        format!(
+                            "HSR ports on interface {} cannot have different \
+                             MAC addresses",
+                            hsr_iface.base.name
+                        ),
+                    ));
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    fn copy_hsr_mac(&mut self) -> Result<(), NmstateError> {
+        let mut pending_changes = HashMap::new();
+
+        for hsr_iface in self.kernel_ifaces.iter().filter_map(|(_, i)| {
+            if !i.is_desired() {
+                return None;
+            }
+
+            if let Interface::Hsr(hsr_iface) = i.desired.as_ref()? {
+                return Some(hsr_iface);
+            }
+
+            None
+        }) {
+            if let Some(hsr_conf) = &hsr_iface.hsr {
+                let mac = if let Some(mac) = &hsr_iface.base.mac_address {
+                    Some(mac.as_ref())
+                } else if let Some(port1) =
+                    self.kernel_ifaces.get(hsr_conf.port1.as_str())
+                {
+                    if let Some(set_mac) =
+                        port1.merged.base_iface().mac_address.as_ref()
+                    {
+                        Some(set_mac.as_ref())
+                    } else {
+                        port1
+                            .merged
+                            .base_iface()
+                            .permanent_mac_address
+                            .as_deref()
+                    }
+                } else {
+                    None
+                };
+
+                if let Some(mac) = mac {
+                    pending_changes
+                        .insert(hsr_iface.base.name.clone(), mac.to_string());
+
+                    pending_changes
+                        .insert(hsr_conf.port1.clone(), mac.to_string());
+
+                    pending_changes
+                        .insert(hsr_conf.port2.clone(), mac.to_string());
+                }
+            }
+        }
+
+        for (iface_name, mac) in pending_changes.drain() {
```

Comment:
No need to `drain()`, with `for ... in pending_changes` will consume pending_changes.